### ISSUE TITLE
Incorrect folder structure with AwsProductRequest and data_folder

### ISSUE THREAD

**Issue by garbagemule** [sentinel-hub/sentinelhub-py#14]
This is related to #12, but since it is a bug (or user error...) in the newly introduced functionality, I felt a new issue would be good.

# Problem

I can't seem to get the new additional folders, `AUX_DATA`, `HTML`, and `rep_info`, to show up in the right place. On the face of it, it looks like the `data_folder` was left out somewhere. Here's a snippet that reproduces the issue:

````python
from sentinelhub.data_request import AwsProductRequest

product_id = 'S2A_MSIL1C_20180110T000231_N0206_R030_T55GCM_20180110T030215'
data_folder = 'work'

product_request = AwsProductRequest(
    product_id=product_id,
    data_folder=data_folder,
    safe_format=True
)
product_request.save_data()
````

And here's the initial folder structure for completeness (the snippet is in `product_test.py`):

````
$ tree
.
└── product_test.py
````

The resulting folder structure looks like this:

````
$ tree -L 3
.
├── S2A_MSIL1C_20180110T000231_N0206_R030_T55GCM_20180110T030215.SAFE
│   ├── AUX_DATA
│   ├── DATASTRIP
│   │   └── DS_SGS__20180110T030215_S20180110T000815
│   ├── HTML
│   └── rep_info
├── product_test.py
└── work
    └── S2A_MSIL1C_20180110T000231_N0206_R030_T55GCM_20180110T030215.SAFE
        ├── DATASTRIP
        ├── GRANULE
        ├── INSPIRE.xml
        ├── MTD_MSIL1C.xml
        └── manifest.safe
````

I eventually found out that there are references to `data_folder` in places where it wouldn't actually have been set yet. This snippet is from `AwsProductRequest`:

````python
    def create_request(self):
        if self.safe_format:
            self.aws_service = SafeProduct(self.product_id, tile_list=self.tile_list, bands=self.bands,
                                           metafiles=self.metafiles)
        else:
            self.aws_service = AwsProduct(self.product_id, tile_list=self.tile_list, bands=self.bands,
                                          metafiles=self.metafiles)

        self.download_list = self.aws_service.get_requests()
````

Note that `self.data_folder` is never passed to the `SafeProduct` constructor (and thus not the `AwsProduct` constructor either). This means that when the call to `get_requests` happens, the variable will be the default of `''`, which propagates down to the `structure_recursion` call in `SafeProduct`. As a result, when creating the folder structure, the data folder is ignored and everything is created in the parent folder. The odd thing is that the files are actually downloaded to the right place, even though the folder structure creation fails.

# Solution (aka 'The Hacky Workaround')

As a temporary fix, I changed the `create_request` method in `AwsProductRequest` so that it passes the data folder as well, and I changed the `structure_recursion` method in `AwsService` so that it joins the data folder with the folder name that is being created:

````python
    def create_request(self):
        if self.safe_format:
            self.aws_service = SafeProduct(self.product_id, tile_list=self.tile_list, bands=self.bands,
                                           #metafiles=self.metafiles)
                                           metafiles=self.metafiles, data_folder=self.data_folder)
````
````python
    def structure_recursion(self, struct, folder):
        if not struct:
            #make_folder(folder)
            make_folder(os.path.join(self.data_folder, folder))
            return
````

This resulted in a nested work folder for some reason, but at least nothing is stored in the parent folder:

````
$ tree -L 3
.
├── product_test.py
└── work
    └── work
        └── S2A_MSIL1C_20180110T000231_N0206_R030_T55GCM_20180110T030215.SAFE
````

To fix the nesting, I changed the call to `structure_recursion` in `SafeProduct` so that it passes the empty string instead of `self.data_folder`:

````python
class SafeProduct(AwsProduct):
    def get_requests(self):
        #self.structure_recursion(safe, self.data_folder)
        self.structure_recursion(safe, '')
````

And voila, the correct folder structure:

````
$ tree -L 3
.
├── product_test.py
└── work
    └── S2A_MSIL1C_20180110T000231_N0206_R030_T55GCM_20180110T030215.SAFE
        ├── AUX_DATA
        ├── DATASTRIP
        ├── GRANULE
        ├── HTML
        ├── INSPIRE.xml
        ├── MTD_MSIL1C.xml
        ├── manifest.safe
        └── rep_info
````

# Conlusion

I haven't actually tested what kind of repercussions my hacks may have introduced, so I am not recommending that these changes be implemented as-is (that's why I didn't make a pull request), but I do hope that they can at least provide a hint as to where to start tearing stuff apart.

Thanks for your time.

**Comment by AleksMat** [comment]
Hi garbagemule,

you are right, the fix of #12 we did 2 days ago wasn't good at all. The `data_folder` parameter in `AwsService` did not have the part of file path given by user.

In the newly released version `1.0.4` this issue should be fixed. We decided to take a bit different approach. The list of empty folders is now returned back to `AwsRequest` class. The folders then get created only upon saving data with either `get_data` or `save_data` method.

Thank you for very nice description of the problem and possible solution. Please let us know if we missed anything else.

**Comment by garbagemule** [comment]
Awesome! Just pulled from pypi and it works. Thanks for a quick turnaround! :)

It isn't a big issue for me because I set the data folder explicitly, but it seems that if you don't, you will now get an error. I made a pull request with a fix (#15) that introduces strong defaults for the data folder variables and arguments in the affected classes. It's only been tested locally (Conda 4.3.30, python 3.5.4, macOS 10.12.6), but the changes are minor :)

Thanks again! Closing this issue as it has been resolved :)
